pkg/controller/nodes/attr_path_resolver.go

```diff
@@ -0,0 +1,149 @@
+
+	for _, attr := range bindAttrPath {
+		switch currVal.GetValue().(type) {
+		case *core.Literal_Map:
+			tmpVal, exist = currVal.GetMap().GetLiterals()[attr.GetStringValue()]
+			if exist == false {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "key [%v] does not exist in literal %v", attr.GetStringValue(), currVal.GetMap().GetLiterals())
+			}
+			currVal = tmpVal
+			count += 1
+		case *core.Literal_Collection:
+			if int(attr.GetIntValue()) >= len(currVal.GetCollection().GetLiterals()) {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "index [%v] is out of range of %v", attr.GetIntValue(), currVal.GetCollection().GetLiterals())
+			}
+			currVal = currVal.GetCollection().GetLiterals()[attr.GetIntValue()]
+			count += 1
+		// scalar is always the leaf, so we can break here
+		case *core.Literal_Scalar:
+			break
+		}
+	}
+
+	// resolve dataclass
+	if currVal.GetScalar() != nil && currVal.GetScalar().GetGeneric() != nil {
+		st := currVal.GetScalar().GetGeneric()
+		// start from index "count"
+		currVal, err = resolveAttrPathInPbStruct(ctx, nodeID, st, bindAttrPath[count:])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return currVal, nil
+}
+
+func resolveAttrPathInPbStruct(ctx context.Context, nodeID string, st *structpb.Struct, bindAttrPath []*core.PromiseAtrribute) (*core.Literal, error) {
+
+	var currVal interface{}
+	var tmpVal interface{}
+	var exist bool
+
+	currVal = st.AsMap()
+
+	// Turn the current value to a map so it can be resolved more easily
+	for _, attr := range bindAttrPath {
+		switch currVal.(type) {
+		// map
+		case map[string]interface{}:
+			tmpVal, exist = currVal.(map[string]interface{})[attr.GetStringValue()]
+			if exist == false {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "key [%v] does not exist in literal %v", attr.GetStringValue(), currVal)
+			}
+			currVal = tmpVal
+		// list
+		case []interface{}:
+			if int(attr.GetIntValue()) >= len(currVal.([]interface{})) {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "index [%v] is out of range of %v", attr.GetIntValue(), currVal)
+			}
+			currVal = currVal.([]interface{})[attr.GetIntValue()]
+		}
+	}
+
+	literal := &core.Literal{}
+
+	// Different types of current value will be wrapped in different ways
+	switch currVal.(type) {
+	// We don't need to wrap map[string]interface{} to LiteralMap, because DictTransformer in flytekit can also handle pb struct
+	case map[string]interface{}:
+		new_st, err := structpb.NewStruct(currVal.(map[string]interface{}))
+		if err != nil {
+			return nil, err
+		}
+		literal.Value = &core.Literal_Scalar{
+			Scalar: &core.Scalar{
+				Value: &core.Scalar_Generic{
+					Generic: new_st,
+				},
+			},
+		}
+	case []interface{}:
+		literals := []*core.Literal{}
+		for _, v := range currVal.([]interface{}) {
+			literalItem := &core.Literal{}
+			scalar, err := resolveInterfaceToLiteralScalar(ctx, nodeID, v)
```

Comment:
does it work if the current value is a nested list? [[1, 2, 3]]

Reply:
brilliant! I made it recursive to handle nested cases and added a test case for that